test/prototype/test_quantized_training.py

```diff
@@ -0,0 +1,225 @@
+    @parametrize("bias", [False, True])
+    @parametrize("device", _DEVICES)
+    def test_int8_linear_compile(self, leading_dims, bias, device):
+        torch._dynamo.reset()
+        embed_dim = 128
+
+        linear_eager = nn.Linear(embed_dim, embed_dim, bias=bias, device=device)
+        quantize_(linear_eager, int8_weight_only_quantized_training())
+        linear_compiled = copy.deepcopy(linear_eager)
+        linear_compiled.compile()
+
+        input_eager = torch.randn(leading_dims + (embed_dim,), device=device) * 10
+        input_compiled = input_eager.clone()
+        input_eager.requires_grad_(True)
+        input_compiled.requires_grad_(True)
+
+        # quantize_() will set torch.set_float32_matmul_precision("high"), which causes segfault.
+        # manually override it here.
+        torch.set_float32_matmul_precision("highest")
+
+        out_eager = linear_eager(input_eager)
+        out_compiled = linear_compiled(input_compiled)
+        torch.testing.assert_close(out_eager, out_compiled)
+
+        grad = torch.randn(leading_dims + (embed_dim,), device=device)
+        out_eager.backward(grad)
+        out_compiled.backward(grad)
+        torch.testing.assert_close(input_eager.grad, input_compiled.grad)
+        torch.testing.assert_close(linear_eager.weight.grad, linear_compiled.weight.grad)
+        if bias:
+            torch.testing.assert_close(linear_eager.bias.grad, linear_compiled.bias.grad)
+
+    @parametrize("compile", [False, True])
+    @parametrize("device", _DEVICES)
+    def test_int8_linear_training(self, compile, device):
+        torch._dynamo.reset()
+        bsize = 4
+        embed_dim = 32
+        n_classes = 10
+
+        model_fp32 = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 2, bias=False),
+            nn.GELU(),
+            nn.Linear(embed_dim * 2, n_classes),
+        ).to(device)
+        model_int8 = copy.deepcopy(model_fp32)
+        quantize_(model_int8, int8_weight_only_quantized_training())
+
+        if compile:
+            model_fp32.compile()
+            model_int8.compile()
+
+        optim_fp32 = AdamW(model_fp32.parameters())
+        optim_int8 = AdamW(model_int8.parameters())
+
+        # prevent segfault with torch.compile()
+        torch.set_float32_matmul_precision("highest")
+
+        for _ in range(5):
+            inputs = torch.randn(bsize, embed_dim, device=device)
+            labels = torch.randint(n_classes, size=(bsize,), device=device)
+            loss_fp32 = F.cross_entropy(model_fp32(inputs), labels)
+            loss_int8 = F.cross_entropy(model_int8(inputs), labels)
+
+            rel_error = abs(loss_int8.item() - loss_fp32.item()) / abs(loss_fp32.item())
+            assert rel_error < 2e-3, rel_error
+
+            loss_fp32.backward()
+            optim_fp32.step()
+            optim_fp32.zero_grad()
+
+            loss_int8.backward()
+            optim_int8.step()
+            optim_int8.zero_grad()
+
+
+class TestFSDP2(FSDPTest):
```

Comment:
it feels like we could probably add this as part of standard test suite, that we can use to sanity check if FSDP is supported for any dtype/tensor subclasses